From ab237ab59b34bc3b4063ab288e9848b3067c1349 Mon Sep 17 00:00:00 2001
From: Kedareswara rao Appana <appana.durga.rao@xilinx.com>
Date: Wed, 23 Sep 2015 13:21:27 +0530
Subject: [PATCH 183/188] net: phy: Use proper duplex settings for SGMII case

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

The duplex setting is not set by the Xilinx phy driver
Which causes issues such that SGMII won't switch rates.
When the duplex is not set it is a value of -1 which then messes
Up the adjust_link() function in the driver as -1 is 0xFFFFFFFF
Which than causes it not to see any link state change.

Reported-by: John Linn <john.linn@xilinx.com>
Signed-off-by: Kedareswara rao Appana <appanad@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 718784a01910710d2bb3c5aaa5374a1d379a96a8)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/phy/xilinx_phy.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/net/phy/xilinx_phy.c b/drivers/net/phy/xilinx_phy.c
index 1445b9a..2f864cc 100644
--- a/drivers/net/phy/xilinx_phy.c
+++ b/drivers/net/phy/xilinx_phy.c
@@ -24,6 +24,7 @@
 #include <linux/xilinx_phy.h>
 
 #define MII_PHY_STATUS_SPD_MASK		0x0C00
+#define MII_PHY_STATUS_FULLDUPLEX	0x1000
 #define MII_PHY_STATUS_1000		0x0800
 #define MII_PHY_STATUS_100		0x0400
 #define XPCSPMA_PHY_CTRL_ISOLATE_DISABLE 0xFBFF
@@ -44,6 +45,11 @@ static int xilinxphy_read_status(struct phy_device *phydev)
 		status = phy_read(phydev, MII_LPA);
 		status = status & MII_PHY_STATUS_SPD_MASK;
 
+		if (status & MII_PHY_STATUS_FULLDUPLEX)
+			phydev->duplex = DUPLEX_FULL;
+		else
+			phydev->duplex = DUPLEX_HALF;
+
 		switch (status) {
 		case MII_PHY_STATUS_1000:
 			phydev->speed = SPEED_1000;
-- 
2.0.2

